File: src/utils/parser.ts
import { Task, Priority, Category } from '../types';

/**
 * Simple regex-based parser for brain dump text.
 * Example: "Gym at 6am high priority, call mom 5pm work"
 */
export function parseBrainDump(text: string, date: string): Task[] {
  // Split by comma or newline
  const lines = text.split(/[,\n]/).filter(line => line.trim().length > 0);
  
  return lines.map(line => {
    const raw = line.trim().toLowerCase();
    
    // Extract Time (e.g., 6am, 5:30 pm, 5.30 am, 6.25 pm)
    const timeRegex = /(\d{1,2}([:.]\d{2})?\s*(am|pm))/gi;
    const timeMatch = raw.match(timeRegex);
    const time = timeMatch ? timeMatch[0] : undefined;
    
    // Extract Priority
    let priority: Priority = 'medium';
    if (raw.includes('high')) priority = 'high';
    else if (raw.includes('low')) priority = 'low';
    
    // Extract Category
    let category: Category = 'personal';
    if (raw.includes('work')) category = 'work';
    else if (raw.includes('health') || raw.includes('gym')) category = 'health';
    
    // Clean Title: Remove the extracted time, priority, and category keywords
    let title = line.trim();
    if (time) title = title.replace(new RegExp(time, 'gi'), '');
    title = title
      .replace(/high priority|medium priority|low priority/gi, '')
      .replace(/\b(high|medium|low|work|personal|health|gym)\b/gi, '')
      .replace(/\s+/g, ' ')
      .trim();

    // Fallback if title becomes empty
    if (!title) title = line.trim();

    return {
      id: Math.random().toString(36).substr(2, 9),
      title,
      time,
      priority,
      category,
      status: 'pending',
      date,
    };
  });
}

File: lib/utils/parser.dart
import '../models/task.dart';
import 'package:uuid/uuid.dart';

class TaskParser {
  static List<Task> parseBrainDump(String text, DateTime date) {
    // Split by comma or new lines
    final lines = text.split(RegExp(r'[,\n]')).where((l) => l.trim().isNotEmpty);
    const uuid = Uuid();

    return lines.map((line) {
      String raw = line.trim().toLowerCase();
      
      // 1. Extract Time (Regex: 6am, 5:30 pm, 5.30 am, 6.25 pm)
      final timeRegex = RegExp(r'(\d{1,2}([:.]\d{2})?\s*(am|pm))');
      final timeMatch = timeRegex.firstMatch(raw);
      String? time = timeMatch?.group(0);

      // 2. Extract Priority
      Priority priority = Priority.medium;
      if (raw.contains('high')) priority = Priority.high;
      else if (raw.contains('low')) priority = Priority.low;

      // 3. Extract Category
      Category category = Category.personal;
      if (raw.contains('work')) category = Category.work;
      else if (raw.contains('health') || raw.contains('gym')) category = Category.health;

      // 4. Clean Title (Remove keywords and time from the original line)
      String title = line.trim();
      if (time != null) title = title.replaceAll(time, '');
      
      title = title
        .replaceAll(RegExp(r'\b(high|medium|low|priority|work|personal|health|gym)\b', caseSensitive: false), '')
        .replaceAll(RegExp(r'\s+'), ' ')
        .trim();

      if (title.isEmpty) title = line.trim();

      return Task(
        id: uuid.v4(),
        title: title,
        time: time,
        priority: priority,
        category: category,
        date: date,
      );
    }).toList();
  }
}
